### ISSUE TITLE
fix compatibility with h5py version 3 and unpin tests

### ISSUE THREAD

**Issue by mathause** [pydata/xarray#4570]
h5py version 3.1 broke our tests. I pinned it to version 2.10 in #4569. We should therefore

* fix the issues
* unpin h5py again

The failures could be related to a change how strings are read: https://docs.h5py.org/en/latest/strings.html I am not sure if this has to be fixed in xarray or in h5necdf. I'd be happy if someone else took this one.

Failed tests:

<details>

```
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_zero_dimensional_variable
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_write_store - As...
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_roundtrip_test_data
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_load - Assertion...
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_dataset_compute
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_roundtrip_object_dtype
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_roundtrip_string_data
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_orthogonal_indexing
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_vectorized_indexing
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_isel_dataarray
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_array_type_after_indexing
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_append_write - A...
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_append_overwrite_values
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_write_groups - A...
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_encoding_kwarg_vlen_string
FAILED xarray/tests/test_backends.py::TestH5NetCDFData::test_compression_encoding
FAILED xarray/tests/test_backends.py::TestH5NetCDFFileObject::test_zero_dimensional_variable
FAILED xarray/tests/test_backends.py::TestH5NetCDFFileObject::test_write_store
FAILED xarray/tests/test_backends.py::TestH5NetCDFFileObject::test_roundtrip_test_data
FAILED xarray/tests/test_backends.py::TestH5NetCDFFileObject::test_load - Ass...
FAILED xarray/tests/test_backends.py::TestH5NetCDFFileObject::test_dataset_compute
FAILED xarray/tests/test_backends.py::TestH5NetCDFFileObject::test_roundtrip_object_dtype
FAILED xarray/tests/test_backends.py::TestH5NetCDFViaDaskData::test_encoding_kwarg_vlen_string
FAILED xarray/tests/test_backends.py::TestH5NetCDFViaDaskData::test_compression_encoding
FAILED xarray/tests/test_distributed.py::test_dask_distributed_netcdf_roundtrip[h5netcdf-NETCDF4]
FAILED xarray/tests/test_distributed.py::test_dask_distributed_read_netcdf_integration_test[h5netcdf-NETCDF4]
```
</details>

**Example failure:**

```python traceback
>  assert_allclose(original, computed)
E  AssertionError: Left and right Dataset objects are not close
E  
E  Differing coordinates:
E  L * dim3     (dim3) <U1 'a' 'b' 'c' 'd' 'e' 'f' 'g' 'h' 'i' 'j'
E  R * dim3     (dim3) object b'a' b'b' b'c' b'd' b'e' b'f' b'g' b'h' b'i' b'j'<
```

**Comment by kmuehlbauer** [comment]
@mathause A quick check running `h5netcdf` testsuit suggest it needs fixing in `h5netcdf`. I'll have a closer look.

**Comment by kmuehlbauer** [comment]
The longer I look at that trying to find the differences the more confused I get. Any hints appreciated. Trying to catch up tomorrow.

**Comment by kmuehlbauer** [comment]
I opened an issue over at h5netcdf (https://github.com/shoyer/h5netcdf/issues/81) with my current evaluations.

**Comment by kmuehlbauer** [comment]
This should possibly be resolved by https://github.com/shoyer/h5netcdf/pull/82. At least the xarray testsuite runs with that `h5netcdf-patch` for all versions of `h5py`.

**Comment by kmuehlbauer** [comment]
I've updated https://github.com/h5netcdf/h5netcdf/pull/82. Review from xarray-folks very much appreciated.

**Comment by kmuehlbauer** [comment]
There is a potential fix in #4893.
